draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -2611,39 +2777,137 @@ to the curve E' isogenous to BLS12-381 G1 is given in {{sswu-map-to-3mod4}}.
+
+6. Choose a mapping following the guidelines in {{choosing-mapping}},
+   and select any required parameters for that mapping.
+
+7. Choose h\_eff to be either the cofactor of E or, if a fast cofactor
+   clearing method is to be used, a value appropriate to that method
+   as discussed in {{cofactor-clearing}}.
+
+8. Construct a Suite ID following the guidelines in {{suiteIDformat}}.
+
+When hashing to an elliptic curve not listed in this section, corresponding
+hash-to-curve suites SHOULD be specified as described in this section.
+
+## Suite ID naming conventions {#suiteIDformat}
+
+Suite IDs MUST be constructed as follows:
+
+    CURVE_ID || "-" || HASH_ID || "-" || MAP_ID || "-" || ENC_VAR || "-"
+
+The fields CURVE\_ID, HASH\_ID, MAP\_ID, and ENC\_VAR are
+ASCII-encoded strings of at most 64 characters each.
+Fields can contain only ASCII characters between 0x21 and 0x7E (inclusive)
+other than hyphen and underscore (i.e., 0x2d, and 0x5f).
+As indicated above, each field (including the last) is followed by a hyphen
+("-", ASCII 0x2d); this helps to ensure that Suite IDs are prefix free.
+
+Fields MUST be chosen as follows:
+
+- CURVE\_ID: a human-readable representation of the target elliptic curve.
+
+- HASH\_ID: a human-readable representation of the expand\_message function
+  and any underlying hash primitives used in hash\_to\_field ({{hashtofield}}).
+  This field MUST be constructed as follows:
+
+        EXP_TAG || ":" || HASH_NAME
+
+  EXP\_TAG indicates the expand\_message variant:
+
+    - "XMD" for expand\_message\_md ({{hashtofield-expand-md}}).
+    - "XOF" for expand\_message\_xof ({{hashtofield-expand-xof}}).
+
+  HASH\_NAME is a human-readable name for the underlying hash primitive.
+  As stated above, hyphens are not allowed. Any hyphens in the name of the
+  hash function SHOULD be replaced with "." (ASCII 0x2e).
+
+  As examples:
+
+    1. For expand\_message\_xof ({{hashtofield-expand-xof}}) with SHAKE-128,
+       HASH\_ID is "XOF:SHAKE.128".
+
+    2. For expand\_message\_md ({{hashtofield-expand-md}}) with SHA3-256,
+       HASH\_ID is "XMD:SHA3.256".
+
+- MAP\_ID: a human-readable representation of the map\_to\_curve function
+  as defined in {{mappings}}. These are defined as follows:
+
+    - "SVDW" for or Shallue and van de Woestijne ({{svdw}}).
+    - "SSWU" for Simplified SWU ({{simple-swu}}, {{simple-swu-AB0}}).
+    - "ELL2" for Elligator 2 ({{elligator2}}, {{ell2edwards}}).
+
+- ENC\_VAR: a string indicating the encoding type and other information.
+  The first two characters of this string indicate whether the suite
+  represents a hash\_to\_curve or an encode\_to\_curve operation
+  ({{roadmap}}), as follows:
+
+    - If ENC\_VAR begins with "RO", the suite uses hash\_to\_curve.
+    - If ENC\_VAR begins with "NU", the suite uses encode\_to\_curve.
+    - ENC\_VAR MUST NOT begin with any other string.
+
+    ENC\_VAR MAY also be used to encode other information used to identify
```

Comment:
I consider we must fix the suiteID name. I am still not convinced on using versions. what is the value added to have  suites XYZv01 and XYZv02?  

If some information is really needed to append in the suiteID, then it should be after the last separator. 
ex: `P256-SHA256-SSWU-RO-<here>`

Reply:
The reason that we require the trailing separator is to ensure that suite IDs are prefix-free.

I wanted to have an "optional" field that people could use to put extra stuff when defining a new suite-ID, and I explicitly described it this way because I wanted to make sure that people *did not* add anything after the trailing separator. (I think I should make this more explicit, though.)

So I think there are three options:

1. leave "optional" field as described here, or
2. add an extra blank field, or
3. don't describe any optional field at all (in which case I worry that people will get it wrong)

Reply:
@chris-wood ?

Reply:
Interesting! I was fine with this text, but on reflection, I'm now leaning towards (a) fixing RO/NU and then (b) adding an extra blank field (with text clarifying its role).

Reply:
I agree with you that the cleanest approach seems to be adding an "extra" field for user-defined info.

The only thing that's slightly unfortunate about this is that it means *all* suite names get longer (well, by one character) in order to make sure that user-defined fields don't break prefix-freeness. So, e.g., we will have to change the suite name to

    P256-XMD:SHA.256-SSWU-RO--

(note: *two* trailing hyphens)

The current version, where we let the user-defined fields live in ENC_VAR, is slightly uglier in principle, but in practice it makes the expected case (no user-defined fields) prettier.

I think I favor optimizing for the expected case and letting the user-defined fields be slightly uglier. For all we no, no one will ever use them... and at least they pay their own cost rather than imposing it on everyone else.

(But again, not to be dramatic... that "cost" is just one extra hyphen, after all...)

Reply:
I think we need a tie breaker! @armfazh?

Reply:
:)

It occurs to me: since this is orthogonal to this PR (this text has been in the document for a while), maybe we can discuss in #213 along with the other Suite-ID--related change?

Reply:
I've added a comment on that issue. Can we move the discussion there so we can discuss in parallel with getting the code updated, etc?

Reply:
I am ok on having two separators at the end, if no user info is added 
P256-XMD:SHA.256-SSWU-RO--

however, one suggestion is to put algorithm variants between parenthesis, so, we ensure the last character is always ')' even if user input is omitted.

(Let's move this thread to 213)